xdsl/dialects/csl/csl_stencil.py

```diff
@@ -139,6 +141,16 @@ def __init__(
         )
 
 
+class ApplyOpHasCanonicalizationPatternsTrait(HasCanonicalisationPatternsTrait):
```

Comment:
Am I correct that this gets automatically applied because `ApplyOpHasCanonicalizationPatternsTrait` trait?

Reply:
Yes, the `canonicalize` pass picks up on this trait, so it can be scheduled anywhere in the pipeline. The GPU pipeline actually has a few `canonicalize` at different points.